core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -677,9 +684,54 @@ public PlanWithProperties visitFilter(FilterNode node, PreferredProperties prefe
         @Override
         public PlanWithProperties visitTableScan(TableScanNode node, PreferredProperties preferredProperties)
         {
+            // attempt tp push preferred partitioning into the table scan
+            var partitioningProperties = preferredProperties.getGlobalProperties().flatMap(PreferredProperties.Global::getPartitioningProperties);
+            if (isUseTableScanNodePartitioning(session) && partitioningProperties.isPresent()) {
+                Optional<TableScanNode> partitionedNode = pushPartitioningIntoTableScan(plannerContext.getMetadata(), session, node, partitioningProperties.get());
+                if (partitionedNode.isPresent()) {
+                    node = partitionedNode.get();
+                }
+            }
             return new PlanWithProperties(node, deriveProperties(node, ImmutableList.of()));
         }
 
+        private Optional<TableScanNode> pushPartitioningIntoTableScan(Metadata metadata, Session session, TableScanNode node, PreferredProperties.PartitioningProperties partitioningProperties)
+        {
+            verify(node.getAssignments().keySet().containsAll(partitioningProperties.getPartitioningColumns()), "TableScanNode does not read all partitioning columns");
+
+            List<ColumnHandle> partitionColumns;
+            if (partitioningProperties.getPartitioning().isPresent()) {
+                Partitioning partitioning = partitioningProperties.getPartitioning().get();
+                // constant partitioning values cannot be pushed into the table scan
+                if (!partitioning.getArguments().stream().allMatch(Partitioning.ArgumentBinding::isVariable)) {
+                    return Optional.empty();
+                }
+                partitionColumns = partitioning.getArguments().stream()
+                        .map(Partitioning.ArgumentBinding::getColumn)
+                        .map(node.getAssignments()::get)
+                        .toList();
+            }
+            else {
+                partitionColumns = partitioningProperties.getPartitioningColumns().stream()
+                        .map(node.getAssignments()::get)
+                        .toList();
+            }
+
+            if (partitionColumns.isEmpty()) {
+                return Optional.empty();
+            }
+
+            return metadata.applyPartitioning(session, node.getTable(), partitioningProperties.getPartitioning().map(Partitioning::getHandle), partitionColumns)
+                    .map(node::withTableHandle)
+                    // Activate the partitioning if it passes the rules defined in DetermineTableScanNodePartitioning
+                    .map(newNode -> DetermineTableScanNodePartitioning.setUseConnectorNodePartitioning(
```

Comment:
Why it doesn't return `node` in case `setUseConnectorNodePartitioning` doesn't set `withUseConnectorNodePartitioning(true)`?

Reply:
I think that would be more complex, but also it does end up capturing the information that we could have used the partitioning, but the optimizer decided not to use it.  I don't think this shows up in the output today, but it could be added.